rhp/v3/contracts.go

```diff
@@ -0,0 +1,118 @@
+package rhp
+
+import (
+	"math/bits"
+
+	"go.sia.tech/core/consensus"
+	"go.sia.tech/core/types"
+)
+
+// ContractRenewalCost returns the cost of renewing a contract for the renter.
+// In other words, this is the amount of money that the renter needs to fund the
+// contract txn with.
+func ContractRenewalCost(cs consensus.State, pt HostPriceTable, fc types.FileContract, minerFee, basePrice types.Currency) types.Currency {
+	return fc.ValidRenterPayout().Add(pt.ContractPrice).Add(minerFee).Add(basePrice).Add(cs.FileContractTax(fc))
+}
+
+// PrepareContractRenewal constructs a contract renewal transaction.
+func PrepareContractRenewal(currentRevision types.FileContractRevision, hostAddress, renterAddress types.Address, renterKey types.PrivateKey, renterPayout, newCollateral types.Currency, hostKey types.PublicKey, pt HostPriceTable, endHeight uint64) (types.FileContract, types.Currency) {
+	hostValidPayout, hostMissedPayout, voidMissedPayout, basePrice := CalculateHostPayouts(currentRevision.FileContract, newCollateral, pt, endHeight)
+
+	return types.FileContract{
+		Filesize:       currentRevision.Filesize,
+		FileMerkleRoot: currentRevision.FileMerkleRoot,
+		WindowStart:    uint64(endHeight),
+		WindowEnd:      uint64(endHeight + pt.WindowSize),
+		Payout:         taxAdjustedPayout(renterPayout.Add(hostValidPayout)),
+		UnlockHash:     currentRevision.UnlockHash,
+		RevisionNumber: 0,
+		ValidProofOutputs: []types.SiacoinOutput{
+			{Value: renterPayout, Address: renterAddress},
+			{Value: hostValidPayout, Address: hostAddress},
+		},
+		MissedProofOutputs: []types.SiacoinOutput{
+			{Value: renterPayout, Address: renterAddress},
+			{Value: hostMissedPayout, Address: hostAddress},
+			{Value: voidMissedPayout, Address: types.Address{}},
+		},
+	}, basePrice
+}
+
+// CalculateHostPayouts calculates the contract payouts for the host.
+func CalculateHostPayouts(fc types.FileContract, newCollateral types.Currency, pt HostPriceTable, endHeight uint64) (types.Currency, types.Currency, types.Currency, types.Currency) {
+	// Calculate the base price and base collateral. The price always includes
+	// the fee for renewing the contract.
+	basePrice := pt.RenewContractCost
+	var baseCollateral types.Currency
+
+	// if the contract height did not increase both prices are zero
+	if contractEnd := uint64(endHeight + pt.WindowSize); contractEnd > fc.WindowEnd {
+		timeExtension := uint64(contractEnd - fc.WindowEnd)
+		basePrice = basePrice.Add(pt.WriteStoreCost.Mul64(fc.Filesize).Mul64(timeExtension))
+		baseCollateral = baseCollateral.Add(pt.CollateralCost.Mul64(fc.Filesize).Mul64(timeExtension))
+	}
+
+	// calculate payouts
+	hostValidPayout := pt.ContractPrice.Add(basePrice).Add(baseCollateral).Add(newCollateral)
+	voidMissedPayout := basePrice.Add(baseCollateral)
```

Comment:
This would cause the host to burn `pt.RenewContractCost`. Is that the behavior in `siad`?

Reply:
It looks like that is the behavior in `siad`.